File: csharp/HiveLLM.Thunder/ClientConfig.cs
namespace HiveLLM.Thunder;

/// <summary>How a <see cref="Credentials"/> value authenticates (CLT-002).</summary>
public enum CredentialKind
{
    /// <summary>Bearer token (<c>token</c> key under HelloMandatory).</summary>
    Token,

    /// <summary>API key (<c>api_key</c> key under HelloMandatory, single-arg <c>AUTH</c> under AuthCommand).</summary>
    ApiKey,

    /// <summary>User + password (<c>AUTH [user, pass]</c> under AuthCommand).</summary>
    UserPass,
}

/// <summary>
/// Credentials for the configured handshake (CLT-002). Auth state is
/// per-connection and sticky — there are no per-call credentials (CLT-003).
/// </summary>
public sealed class Credentials
{
    private Credentials(CredentialKind kind, string secret, string? user)
    {
        Kind = kind;
        Secret = secret;
        User = user;
    }

    /// <summary>Which credential style this is.</summary>
    public CredentialKind Kind { get; }

    internal string Secret { get; }

    internal string? User { get; }

    /// <summary>Bearer token (<c>token</c> key under HelloMandatory).</summary>
    public static Credentials Token(string token)
    {
        ArgumentNullException.ThrowIfNull(token);
        return new Credentials(CredentialKind.Token, token, null);
    }

    /// <summary>API key (<c>api_key</c> under HelloMandatory, single-arg <c>AUTH</c> under AuthCommand).</summary>
    public static Credentials ApiKey(string apiKey)
    {
        ArgumentNullException.ThrowIfNull(apiKey);
        return new Credentials(CredentialKind.ApiKey, apiKey, null);
    }

    /// <summary>User + password (<c>AUTH [user, pass]</c> under AuthCommand configs).</summary>
    public static Credentials UserPass(string user, string pass)
    {
        ArgumentNullException.ThrowIfNull(user);
        ArgumentNullException.ThrowIfNull(pass);
        return new Credentials(CredentialKind.UserPass, pass, user);
    }
}

/// <summary>
/// Client configuration: connect timeout default <b>10 s</b> (CLT-001),
/// per-call timeout default <b>30 s</b> (CLT-020), optional credentials and
/// client name for the handshake (CLT-002).
/// </summary>
public sealed record ClientConfig
{
    /// <summary>TCP connect timeout (CLT-001). Default 10 s.</summary>
    public TimeSpan ConnectTimeout { get; init; } = TimeSpan.FromSeconds(10);

    /// <summary>
    /// Default per-call timeout (CLT-020); override per call with the
    /// timeout-taking <see cref="ThunderClient.CallAsync(string, IReadOnlyList{Value}, TimeSpan, CancellationToken)"/>
    /// overload. Default 30 s.
    /// </summary>
    public TimeSpan CallTimeout { get; init; } = TimeSpan.FromSeconds(30);

    /// <summary>Handshake credentials, when the protocol config wants them.</summary>
    public Credentials? Credentials { get; init; }

    /// <summary>Client identifier sent in the <c>HELLO</c> map (HelloMandatory). Default <c>thunder-client</c>.</summary>
    public string? ClientName { get; init; }

    /// <summary>
    /// Optional client TLS (FR-29 / SPEC-008 CAN-020), <b>off by default</b>.
    /// Non-null makes the client complete a
    /// <see cref="System.Net.Security.SslStream"/> handshake before any Thunder
    /// frame; null keeps plaintext, the default path, entirely unchanged. There
    /// is no STARTTLS — TLS is decided at connect time. A TLS setup, handshake
    /// or certificate-verification failure classifies as
    /// <see cref="ThunderConnectionException"/>, exactly like a plaintext
    /// connect failure.
    /// </summary>
    public ClientTls? Tls { get; init; }
}

/// <summary>
/// Client-side TLS material (FR-29 / SPEC-008 CAN-020). Presence of this on
/// <see cref="ClientConfig.Tls"/> makes the client dial TLS; absence keeps it
/// plaintext (the default). Mirrors the Rust <c>ClientTls</c> — plain data that
/// always compiles regardless of whether TLS is used.
/// </summary>
public sealed record ClientTls
{
    /// <summary>
    /// Name to verify the server certificate against (the TLS target host /
    /// SNI). When null, the endpoint host is used.
    /// </summary>
    public string? ServerName { get; init; }

    /// <summary>
    /// Path to a PEM file of trusted root(s) to pin. When null, the platform's
    /// native root store is used.
    /// </summary>
    public string? CaPath { get; init; }
}

/// <summary>What the handshake learned about this connection (CLT-002).</summary>
/// <param name="Authenticated">
/// True once the server accepted the credentials (<c>AUTH</c> succeeded or
/// the <c>HELLO</c> reply said so).
/// </param>
/// <param name="Capabilities">Capability names from the <c>HELLO</c> reply (HelloMandatory).</param>
public sealed record HandshakeInfo(bool Authenticated, IReadOnlyList<string> Capabilities)
{
    /// <summary>Unauthenticated, no capabilities.</summary>
    public static HandshakeInfo Default { get; } = new(false, System.Array.Empty<string>());
}

File: python/thunder_rpc/client_config.py
"""Client configuration shared by the sync and async clients (SPEC-003).

Distinct from :class:`~thunder_rpc.config.Config`, which describes the
*protocol* one application speaks (SPEC-002): this is the per-client dialing
policy — timeouts, credentials, client name.
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class Credentials:
    """Credentials for the handshake (CLT-002). Auth state is
    per-connection and sticky — there are no per-call credentials (CLT-003).

    Build via the factories: :meth:`token`, :meth:`api_key`,
    :meth:`user_pass`.
    """

    #: ``"token"`` | ``"api_key"`` | ``"user_pass"``.
    kind: str
    #: The secret parts: ``(token,)``, ``(api_key,)`` or ``(user, password)``.
    secrets: tuple[str, ...]

    TOKEN = "token"
    API_KEY = "api_key"
    USER_PASS = "user_pass"

    @classmethod
    def token(cls, token: str) -> Credentials:
        """Bearer token (``token`` key under ``HELLO_MANDATORY``)."""
        return cls(cls.TOKEN, (token,))

    @classmethod
    def api_key(cls, api_key: str) -> Credentials:
        """API key (``api_key`` key under ``HELLO_MANDATORY``, single-arg
        ``AUTH`` under ``AUTH_COMMAND``)."""
        return cls(cls.API_KEY, (api_key,))

    @classmethod
    def user_pass(cls, user: str, password: str) -> Credentials:
        """User + password (``AUTH [user, pass]`` under ``AUTH_COMMAND``)."""
        return cls(cls.USER_PASS, (user, password))


@dataclass(frozen=True)
class ClientConfig:
    """Client configuration: connect timeout default **10 s** (CLT-001),
    per-call timeout default **30 s** (CLT-020), optional credentials and
    client name for the handshake (CLT-002).
    """

    #: TCP connect timeout in seconds (CLT-001).
    connect_timeout: float = 10.0
    #: Default per-call timeout in seconds (CLT-020); override per call with
    #: ``call(..., timeout=...)``.
    call_timeout: float = 30.0
    #: Handshake credentials, when the protocol config wants them.
    credentials: Credentials | None = None
    #: Client identifier sent in the ``HELLO`` map (``HELLO_MANDATORY``).
    client_name: str | None = None


@dataclass(frozen=True)
class HandshakeInfo:
    """What the handshake learned about this connection (CLT-002)."""

    #: ``True`` once the server accepted the credentials (``AUTH`` succeeded
    #: or the ``HELLO`` reply said so).
    authenticated: bool = False
    #: Capability names from the ``HELLO`` reply (``HELLO_MANDATORY``).
    capabilities: tuple[str, ...] = ()
